### DISCUSSION TITLE
Does EPOCH control the energy balance?

### DISCUSSION THREAD

**Discussion by DanRRRR** [epochpic/epoch#418] [General]
The major quality characteristics of any PIC or hydro codes is energy balance control. Does EPOCH control at any time how energy was conserved in the system? That means what laser energy is currently in the box, what was absorbed and reflected, what energy is in electrons, ions, what energy was spent on atom ionization to reach the current state Z, and what energy have left the simulation box in all kinds of particles and fields?

**Comment by Status-Mirror** [comment]
Hey Dan,

The core PIC algorithm is not energy conserving. PIC codes are subject to numerical heating, which artificially increases particle energies as the simulation progresses. It is up to the user to choose a particle-per-cell count and cell-size to minimise this heating to an acceptable level. This minimum resolution can be quickly studied by running a small scale simulation to for the required runtime, with just the plasma present. The final temperature gives you a self heating estimate.

Ionisation routines conserve energy when averaged over many particles.

The user can track particles which leave the simulation window, and the field energy which escapes, using the output block keys:
ejected_particles
absorption 

Hope this helps
Stuart

**Comment by DanRRRR** [comment]
Thanks Stuart, it always very nice to get your quick and helpful responds. Would also be great to have many users participating as this  is great code which really deserves attention. Even my initial and very limited experience with it very pleasantly amazed me. 

Yes, this is not only the PIC codes problem but the problem of many other methods. But taking that in mind having the facility controlling how far energy balance is off would be still very helpful. Sometimes the energy balance is not that bad and then abruptly gets exploded while you are keeping running without noticing it. 

With this respect, can the code similarly with lost the box particles assess the reflected/scattered laser energy ?

**Reply by Status-Mirror** [reply]
Hey Dan,

I don't think EPOCH currently has this capability, but it would be possible to calculate global simulation energy. The issue is, how would you fix it?

You could artificially decrease particle energies, but this wouldn't produce the physically correct answer. Self heating (as I understand it), comes from simulation noise, and won't affect all particles equally. Any attempt to correct this trades one inaccurate result for another.

Currently EPOCH can model injected and escaping field energy. Provided self-healing is minimal, you can use this to estimate energy in and energy out. It would be good to calculate this by boundary - another user has requested  boundary-specific escaping energy. We can look into this.

In the meantime, your best approach may be to study the Pointing flux. This is a grid variable which gives you the laser intensity in each direction. Tracking this near the boundaries over the simulation runtime can tell you which boundaries the laser energy moves towards.

Hope this helps,
Stuart

**Comment by DanRRRR** [comment]
Often you even have no clue that artificial heating happens. You tend to attribute it to some valid physical effects. And only over many lost runs you will notice something fishy. With experience with some other PIC codes this happened if some initial conditions were set not right. For example the time step was chosen too large or too low. I'm new with EPOCH and do not know how this is set here and is it possible to control and correct it over the course of run). Potentially, assessing and controlling this effect would be top quality assurance feature i suspect no other PIC codes currently have.
Great if EPOCH calculates laser reflected light. The #1 fashion feature would be if it gets also its angular dependence and harmonics :)
